cli/remotebazel/remotebazel.go

```diff
@@ -167,27 +169,67 @@ func determineRemote(repo *git.Repository) (*git.Remote, error) {
 }
 
 func determineDefaultBranch(repo *git.Repository) (string, error) {
-	branches, err := repo.Branches()
+	f := sync.OnceValues(func() (string, error) {
+		branch, localErr := localDefaultRef(repo)
+		if localErr == nil {
+			return branch, nil
+		}
+		branch, remoteErr := remoteDefaultRef(repo)
+		if remoteErr == nil {
+			return branch, nil
+		}
+
+		return "", status.NotFoundErrorf("could not determine default branch: %s", errors.Join(localErr, remoteErr))
+	})
+
+	return f()
+}
+
+func localDefaultRef(repo *git.Repository) (string, error) {
+	var localErr error
+	for _, defaultRefName := range defaultRefNames {
+		_, err := repo.Reference(plumbing.ReferenceName(defaultRefName), false /*resolved*/)
+		if err != nil {
+			localErr = errors.Join(localErr, err)
+			continue
+		}
+		return defaultRefName, nil
+	}
+
+	return "", status.NotFoundErrorf("could not determine local default branch: %s", localErr)
+}
+
+func remoteDefaultRef(repo *git.Repository) (string, error) {
```

Comment:
I think we should do something like https://stackoverflow.com/a/50056710 so that we aren't depending on the branches being named "main" or "master"

Also, it looks like `getBaseBranchAndCommit` is already doing a `git ls-remote origin` call, unconditionally. Maybe we can instead parse the output from `git remote show`, so that we don't have to do 2 remote fetches. (These have relatively high latency, like 400 ms for each call, so we should try to just do one if possible)

Reply:
We can try to read the symbolic ref to determine the default branch as part of the `local` strategy.

```
git symbolic-ref refs/remotes/origin/HEAD
```

However, as I noted in the PR summary, in GitHub Actions `actions/checkout` plugin does not write `refs/remotes/origin/HEAD` symbolic ref. So in such a case, we would still need to fall back onto a remote-based option (with `git ls-remote`).

Note that `git remote show` is a higher level command that does a combination of reading the remotes/origin/HEAD ref and calling `git ls-remote`. `git remote show -n` would skip the `ls-remote` call, but the underlying file-parsing methodology still applies.

Reply:
I made some changes to attempt to find the local `refs/remotes/origin/*` refs before finally reaching out to `git ls-remote` as the last resort.

Also added a warning log there just to make sure.

Let me know what you think.

Reply:
Because the github checkout action doesn't set the local state, remote bazel will probably end up making the remote call pretty frequently. I think Brandon's point is that now we'll make duplicate calls to `ls-remote` in `getBaseBranchAndCommit` and here `originRemote.Fetch(&git.FetchOptions{}); refs, err := originRemote.List(&git.ListOptions{})` Could we make the call once and pass into each function?

Reply:
Let me look into this.
Trickle-down the refs should be fine.

But I feel like we could use a dedicated `//cli/git` package with our own `repo` struct that caches all these results to re-use within an invocation. 🤔 
Something for a future PR.